### PR TITLE
[api-extractor] Fix #3593 Incorrect canonical reference to aliased class in .api.json

### LINKED ISSUE #3593: [api-extractor] Incorrect canonical reference to aliased class in .api.json
## Summary

While waiting for #1596 to be implemented, I tried to find a workaround to generate documentation for our packages that have multiple entry-points under different paths, but did not succeed caused by what I consider bugs in api-extractor's canonical reference resolving.

For packages with multiple entry points, which api-extractor currently does not support, [this comment](https://github.com/microsoft/rushstack/issues/664#issuecomment-390443043) suggests the following approach:

> If you don't need rollups, as a workaround you can simply create a fake entry point that reexports all your real entry points. That will work reasonably well for the API report and documentation features.

The "legacy" public API we want to generate documentation for uses one file per class and groups classes into subdirectories. To keep the original structure and to avoid name clashes, we must somehow retain the path name in the fake entry point.

To achieve this, I tried two approaches for the "fake" entry point `index.d.ts`:

1. Alias the classes with a path prefix, like so:
   `export { Foo as subpath_Foo } from "./subpath/Foo"`
2. Re-export only direct members and convert sub-paths to namespaces with their own `index.d.ts`, which can then be re-exported like so:
   `import * as subpath from "./subpath"; export { subpath }`
   (this is the suggested workaround for `export * as subpath from "./subpath"` not yet being supported by api-extractor)

Both approaches lead to incorrect references from a subclass to its (aliased) superclass.

## Repro steps

To reproduce, clone the [example repo](https://github.com/fwienber/api-extractor-reference-bug.git) and run

```
api-extractor run --local
```
The example project contains a `src` folder with four TypeScript `.d.ts` files with one class declaration each: `A`, `B`, `somepath/A` and `somepath/B`. In both cases, `B` is a subclass of `A`.
There is a "fake" entry point file `src/index.d.ts`, which re-exports all four entry points. It implements both approaches described above, namely using aliases and using a namespace, for which there is another "fake" sub-entry point `src/somepath/index.d.ts` that re-exports `A` and `B` from under the `somepath` path.

  **Expected result:** 

The generated file `input/api-extractor-test.api.json` should contain the correct canonical reference from namespaced module `somepath.B` to its superclass `somepath.A`:
```
    "canonicalReference": "api-extractor-test!somepath.A:class"
```
and/or the correct canonical reference from aliased module `somepath_B` to its superclass `somepath_A`:
```
    "canonicalReference": "api-extractor-test!somepath_A:class"
```

  **Actual result:** 

For both approaches, the reference incorrectly points to the top-level `A`:

```
    "canonicalReference": "api-extractor-test!A:class"
```

If you find the JSON file too confusing, you can generate Markdown files simply by calling
```
api-documenter markdown
```
The resulting Markdown files in folder `markdown` contain the corresponding wrong hyperlinks in file `api-extractor-test.somepath.b.md` (`[A](./api-extractor-test.a.md)` instead of `[A](./api-extractor-test.somepath.a.md)`) and `api-extractor-test.somepath_b.md` (`[A](./api-extractor-test.a.md)` instead of `[A](./api-extractor-test.somepath_a.md)`).
Of course, this is not api-documenter's fault, as the `*.api.json` file already contains the wrong canonical references.

## Details

It seems that neither alias support (`export { __ as __ }`) nor namespace aliases (`import * as __`) are taken into account when resolving references.

The name-clash between `A` and `somepath/A` has been used to motivate the need for aliasing, but even without an actual clash, it still does not work.
If you change the name of `somepath/A` to something else in the example workspace, the superclass reference of `somepath/B` in api-extractor's output points to an incorrect, non-existing module name (also missing the path-part of the actual module name) and thus no hyperlink at all is rendered by api-documenter.

## Standard questions

Please answer these questions to help us investigate your issue more quickly:

| Question | Answer |
| -------- | -------- |
| `@microsoft/api-extractor` version? | 7.29.2 |
| Operating system? | Linux |
| API Extractor scenario? | docs (.api.json) |
| Would you consider contributing a PR? | Maybe (need help) |
| TypeScript compiler version? | 4.7.4 |
| Node.js version (`node -v`)? | v16.15.1 |

### PR DESCRIPTION
**Pull request by fwienber**
## Summary

api-extractor currently does not support packages with multiple entry points.
As a workaround, multiple entry points can be aggregated through a single entry point, directly or indirectly re-exporting them.
This works as long as their original names are used. However, when trying to rename exports, for example because they are grouped in directories and when "flattening", name-clashes must be avoided, api-extractor produces document model (`*.api.json`) files with incorrect canonical references. Both renaming and moving a module into a namespace do not work:

1. Renaming: `export { Foo as subpath_Foo } from "./subpath/Foo"`
2. Namespacing: `import * as subpath from "./subpath"; export { subpath }` (this is the workaround api-extractor suggests for `export * as subpath from "./subpath"` not yet being supported)

In both cases, the `subpath` information is lost in canonical references to `Foo`.
For details and an example, see #3593.

Fixes #3593

## Details

After a pointer by @zelliott, extended `Collector` to offer a new function to retrieve the "emit name" of a given `ts.Symbol`. The `emitName` property can be fetched from a `CollectorEntity` and is rename-aware. `DeclarationReferenceGenerator` calls this function to use the correct emit name if the referenced symbol has been renamed via a re-export. This fixes the first part of the bug.

Concerning the second part of the bug, namespaced re-exports:
When `DeclarationReferenceGenerator#_getParentReference()` looks for a parent symbol, it must also resolve namespace parent symbols the Collector derived from `import * as ___`. To that end, `Collector` now has a function `getExportingNamespace()` that finds namespace imports of the given symbol by looking up the parent namespace in the corresponding `CollectorEntity` (if present). 
This new function is then used by `DeclarationReferenceGenerator` to compute the correct parent `DeclarationReference`. Also there, handling of namespaces had to be improved.

## How it was tested

Added two test scenarios, one for each part of the bug:
`docReferencesAlias` uses renaming, `docReferencesNamespaceAlias` uses namespaces for re-export.
When building the `build-tests`, there is no git diff, so the expected `*.api.json` files with the correct canonical references are now generated.

### CODE REVIEW

**Review by zelliott — COMMENTED** [review]
Thanks for the draft PR. Left some initial comments. I need to dive into this deeper, hoping to get to that this weekend. Test cases look good. 👍

File: apps/api-extractor/src/collector/Collector.ts
Comment (fwienber) [reply apps/api-extractor/src/collector/Collector.ts:341]: Yes, this is the code I was referring to somewhere in my length commit message (should probably have been in the code):

> Namespace imports are now also registered in the
_entitiesBySymbol Map, but I did not find the API to
access the symbol of the AstNamespaceImport astEntity
(ugly type assertion 'as unknown as ts.Type', which is
probably not even the right type to access 'symbol').
I found the reference while debugging, but could not
find it in any meaningful subtype.

I also remember seeing a `moduleSymbol` during debugging, but iirc, it did not point to the desired symbol.
Did you find out where `symbol` comes from? Is it maybe meant to be "private"? Or is there maybe a gap in the type definition of the AST declaration data structure?

File: apps/api-extractor/src/generators/DeclarationReferenceGenerator.ts
Comment (fwienber) [reply apps/api-extractor/src/generators/DeclarationReferenceGenerator.ts:21]: No problem, I can merge-in your branch(es). If there are any complicated conflicts / vanished API, I may have to ask for help.

File: apps/api-extractor/src/collector/Collector.ts
Comment (fwienber) [reply apps/api-extractor/src/collector/Collector.ts:339]: Thanks for the advance notice!

Comment (fwienber) [reply apps/api-extractor/src/collector/Collector.ts:339]: To be ready to update this PR when your PRs are merged, I did the following:
* created an [integration branch for your two PRs](https://github.com/fwienber/rushstack/commits/zelliott/integrate)
* rebased and minimized my changes on top of that branch as a [new branch](https://github.com/fwienber/rushstack/commits/fwienber/api-extractor-fix-3593-reference-alias-rebased)
* created a [draft PR](https://github.com/fwienber/rushstack/pull/1) inside my fork so that you can review the diff to your PRs

Once your PRs are merged, it should be easy to rebase those changes onto main.

File: apps/api-extractor/src/generators/DeclarationReferenceGenerator.ts
Comment (fwienber) [reply apps/api-extractor/src/generators/DeclarationReferenceGenerator.ts:21]: See my comment below for the other conflicting branch.

File: apps/api-extractor/src/collector/Collector.ts
Comment (fwienber) [reply apps/api-extractor/src/collector/Collector.ts:341]: I think this is the only remaining open issue which is still present in my rebased branch. So it would be great if you helped me find a better way to retrieve this namespace declaration symbol.

File: apps/api-extractor/src/collector/Collector.ts
Comment (fwienber) [reply apps/api-extractor/src/collector/Collector.ts:339]: I now also added commits that lead to a similar state to this PR.

File: apps/api-extractor/src/generators/DeclarationReferenceGenerator.ts
Comment (fwienber) [reply apps/api-extractor/src/generators/DeclarationReferenceGenerator.ts]: > What about moving this conversion to a private static utility method inside the class?

I now did that (commit https://github.com/microsoft/rushstack/pull/3602/commits/dd5594a87445370a70f6102c10cfc5e07eb823ae).

**Comment by zelliott** [comment]
Hey, sorry for the delay @fwienber! Hoping to take a look tomorrow 👍

**Comment by fwienber** [comment]
No worries! Looks like we live in time zones that unfortunately have minimal working hours overlap (CEST = UTC+2 here)...

**Review by zelliott — CHANGES_REQUESTED** [review]
I think the `nameForEmit` part LGTM, left a few comments about the namespace imports piece. LMK if I can help clarify anything!

File: apps/api-extractor/src/generators/DeclarationReferenceGenerator.ts
Comment (fwienber) [reply apps/api-extractor/src/generators/DeclarationReferenceGenerator.ts]: Sounds totally convincing. I also felt it was weird to _insert_ the namespace path in retrospect.
I just could not figure out how to find the namespace from the source file (step 4). So I'll just try that again!

File: apps/api-extractor/src/collector/Collector.ts
Comment (fwienber) [reply apps/api-extractor/src/collector/Collector.ts]: Resolved by my last commit https://github.com/microsoft/rushstack/pull/3602/commits/032e4014add5939675e5a1be1b96286c2d08099c.

File: apps/api-extractor/src/generators/DeclarationReferenceGenerator.ts
Comment (fwienber) [reply apps/api-extractor/src/generators/DeclarationReferenceGenerator.ts]: Please have a look at my latest commit https://github.com/microsoft/rushstack/pull/3602/commits/032e4014add5939675e5a1be1b96286c2d08099c. I think that solves it.

**Comment by fwienber** [comment]
> @octogonz JFYI: This PR is will have merge conflicts with #3584 and #3552 and will be a bit easier to implement and review once those two PRs are submitted.

Yes, would be great it @zelliott's two branches got merged. I already have prepared a cleaned-up version of my changes based on microsoft:main + zelliott:refs + zelliott:exported which can then easily be rebased on main.

File: apps/api-extractor/src/generators/DeclarationReferenceGenerator.ts
Comment (fwienber) [reply apps/api-extractor/src/generators/DeclarationReferenceGenerator.ts]: After rebasing, now https://github.com/fwienber/rushstack/commit/ef387a79b49beec1f6bdd444d43c80e08400356d.

**Comment by fwienber** [comment]
> > @octogonz JFYI: This PR is will have merge conflicts with #3584 and #3552 and will be a bit easier to implement and review once those two PRs are submitted.
> 
> Great, looks like they are already merged! Then, I'll force-push this branch with my rebased latest solution!

Done, ready for review!

**Comment by fwienber** [comment]
bump

**Comment by zelliott** [comment]
Hey @fwienber - I opened https://github.com/fwienber/rushstack/pull/2 for some proposed changes to this PR. LMK what you think - feel free to merge in/modify as you see fit.

I don't actually have merge powers in this repository - I'm just a contributor - we'll need @octogonz for that. We'll also definitely want him to take a look at this PR... `AstNamespaceImports` are tricky.

**Comment by fwienber** [comment]
Okay, just go to `build-tests/api-extractor-scenarios/src/docReferencesNamespaceAlias/Item.ts` and add `{@inheritDoc renamed.Options}` to the TSDoc comment of class `Item`.
Unfortunately, when building the api-extractor-scenarios, all unresolved references messages are swallowed, but setting a breakpoint in the line I posted above, where the error is thrown, and then debug the `build` script of `build-tests/api-extractor-scenarios/package.json` should do the job.
Then, `rootAstEntity instanceof AstSymbol` is false because it is the `AstNamespaceImport` with the name `renamed`, so it is exactly what we want, but `AstReferenceResolver` cannot handle it.

File: build-tests/api-extractor-scenarios/etc/exportImportStarAs2/api-extractor-scenarios.api.json
Comment (zelliott) [inline build-tests/api-extractor-scenarios/etc/exportImportStarAs2/api-extractor-scenarios.api.json:196]: This reference is incorrect, but I'm not sure what the right reference is (another instance of https://github.com/microsoft/rushstack/issues/1308). I'm thinking it should probably be `api-extractor-scenarios!~ForgottenClass:class`, and `api-extractor-scenarios!~ForgottenClass:class` is definitely _more correct_ that `"api-extractor-scenarios!ForgottenClass:class"` (as `ForgottenClass` is un-exported from the package, and the `~` indicates that).

I'll leave a comment above for how we can slightly modify `DeclarationReferenceGenerator` to fix this, then feel free to mark this as resolved. 👍

File: apps/api-extractor/src/generators/DeclarationReferenceGenerator.ts
Comment (zelliott) [inline apps/api-extractor/src/generators/DeclarationReferenceGenerator.ts:118]: This is the logic I suggested, but it's not quite right. I think we want:

```
// Otherwise, this symbol is from the current package. If we've found an associated consumable
// `CollectorEntity`, then use Exports. We use `consumable` here instead of `exported` because
// if the symbol is exported from a non-consumable `AstNamespaceImport`, we don't want to use
// Exports. We should use Locals instead.
const entity: CollectorEntity | undefined = this._collector.tryGetEntityForSymbol(symbol);
if (entity?.consumable) {
  return Navigation.Exports;
}
```

After changing this, run `rush rebuild` to see the appropriate changes propagated to api-extractor-scenarios.

**Review by zelliott — APPROVED** [review]
I think this looks good to me! Regarding the `@link` and `@inheritDoc` stuff - I'll take a look, but I don't see it as blocking this PR.

File: apps/api-extractor/src/generators/DeclarationReferenceGenerator.ts
Comment (fwienber) [reply apps/api-extractor/src/generators/DeclarationReferenceGenerator.ts]: done

**Comment by octogonz** [comment]
@fwienber I've reverted your `.gitignore` changes and moved them to a separate PR https://github.com/microsoft/rushstack/pull/3622

File: apps/api-extractor/src/collector/CollectorEntity.ts
Comment (octogonz) [inline apps/api-extractor/src/collector/CollectorEntity.ts:182]: @fwienber I've converted this property to be a method because it does nontrivial work (iterating over a map).  

For example, someone might be tempted to write code like this:

```ts
if (entity.firstExportingConsumableParent?.exported 
  || entity.firstExportingConsumableParent?.consumable) {
}
```

...whereas denoting it as a function call makes the performance concern more conspicuous:

```ts
if (entity.getFirstExportingConsumableParent()?.exported 
  || entity.getFirstExportingConsumableParent()?.consumable) {
}
```

File: common/changes/@microsoft/api-extractor/fwienber-api-extractor-fix-3593-reference-alias_2022-08-29-07-53.json
Comment (octogonz) [inline common/changes/@microsoft/api-extractor/fwienber-api-extractor-fix-3593-reference-alias_2022-08-29-07-53.json:6]: Although this is technically a bug fix, I've converted it to a "minor" version bump because **DeclarationReferenceGenerator.ts** has lots of edge cases that are difficult to test comprehensively.

**Review by octogonz — APPROVED** [review]

File: apps/api-extractor/src/generators/DeclarationReferenceGenerator.ts
Comment (octogonz) [reply apps/api-extractor/src/generators/DeclarationReferenceGenerator.ts]: > One idea is to choose the shortest ID, and among IDs of the same length, choose the one that comes first alphabetically.

This sounds good to me.  Maybe we could also provide a TSDoc tag that allows the choice to be explicitly specified, similar to suggestions for the other [ae-ambiguous-ids](https://github.com/microsoft/rushstack/labels/ae-ambiguous-ids) issues.

**Comment by octogonz** [comment]
@fwienber Thanks for contributing this fix! And thanks @zelliott for reviewing the code!

🚀 Released with API Extractor 7.31.0

**Pull request merged**
